proposals/0000-sum-to-tag.md

```diff
@@ -0,0 +1,259 @@
+---
+author: David Feuer
+date-accepted: ""
+ticket-url: ""
+implemented: ""
+---
+
+This proposal is [discussed at this pull request](https://github.com/ghc-proposals/ghc-proposals/pull/369).
+
+# Add `sumToTag#` primop
+
+Add a `sumToTag#` primop indicating which sum alternative is
+represented by a particular unboxed sum value. For example,
+
+```haskell
+sumToTag# (# a | #) = 0#
+sumToTag# (# | | b #) = 2#
+```
+
+## Motivation
+
+It's sometimes useful to get the numerical value of an alternative.
+For example, I might write
+
+```haskell
+newtype Trit# = Trit# (# (##) | (##) | (##) #)
+
+tritToInt# :: Trit# -> Int#
+tritToInt# (Trit# (# (##)|| #)) = 0#
+tritToInt# (Trit# (# |(##)| #)) = 1#
+tritToInt# (Trit# (# ||(##) #)) = 2#
+```
+
+Under the hood, unboxed sums are actually represented by `Int#` tags
+alongside the "payload" values, often with some padding. For example,
+`(# (##) | Bool #)` is represented as something like `(# Int#, Bool #)`.
+The values look like this:
+
+```haskell
+(# (##) | #)   ~=   (# 1#, padding #)
+(# | b #)      ~=   (# 2#, b #)
+```
+
+Note: the choice to make the first alternative `1#` rather than `0#`
+is poorly motivated, and should probably be changed, but that's beyond
+the strict scope of this proposal.
```

Comment:
Is the one-indexed tag for unboxed sums currently exposed in any way? I agree that zero-indexing makes more sense, but it could be pretty confusing if the actual representation were already exposed.

Reply:
No, it's not. I strongly suspect it should be changed to 0-indexed internally, but that's out of the scope of the proposal process.